full-stack-tests/presentation/src/frontend/Content.test.ts

```diff
@@ -273,12 +274,97 @@ describe("Content", () => {
         className: "Generic:PhysicalObject",
         id: Id64.invalid,
       }]);
-      await validatePagedDistinctValuesResponse(ruleset, typeTwoKey, descriptor, field.getFieldDescriptor(), [{
+      await validatePagedDistinctValuesResponse(imodel, ruleset, typeTwoKey, descriptor, field.getFieldDescriptor(), [{
         displayValue: "",
         groupedRawValues: [undefined],
       }]);
     });
 
+    it("gets distinct content values based on hierarchy level descriptor", async function () {
+      // create an imodel with Model -> Elements relationship
+      const testIModel = await buildTestIModelConnection(this.test!.fullTitle(), async (db) => {
+        const categoryKey = insertSpatialCategory(db, "Category");
+        const modelKeyA = insertPhysicalModel(db, "Model A");
+        insertPhysicalElement(db, "Element A1", modelKeyA.id, categoryKey.id);
+        insertPhysicalElement(db, "Element A2", modelKeyA.id, categoryKey.id);
+        const modelKeyB = insertPhysicalModel(db, "Model B");
+        insertPhysicalElement(db, "Element B", modelKeyB.id, categoryKey.id);
+        insertPhysicalElement(db, "Element B", modelKeyB.id, categoryKey.id);
+      });
+
+      // set up ruleset
+      const ruleset: Ruleset = {
+        id: Guid.createValue(),
+        rules: [{
+          ruleType: RuleTypes.RootNodes,
+          specifications: [{
+            specType: ChildNodeSpecificationTypes.InstanceNodesOfSpecificClasses,
+            classes: [{
+              schemaName: "BisCore",
+              classNames: ["PhysicalModel"],
+              arePolymorphic: false,
+            }],
+            groupByClass: false,
+            groupByLabel: false,
+          }],
+        }, {
+          ruleType: RuleTypes.ChildNodes,
+          condition: `ParentNode.IsOfClass("Model", "BisCore")`,
+          specifications: [
+            {
+              specType: ChildNodeSpecificationTypes.RelatedInstanceNodes,
+              relationshipPaths: [
+                {
+                  relationship: { schemaName: "BisCore", className: "ModelContainsElements" },
+                  direction: "Forward",
+                },
+              ],
+              groupByClass: false,
+              groupByLabel: false,
+            },
+          ],
+        }],
+      };
+      const rootNodes = await Presentation.presentation.getNodes({ imodel: testIModel, rulesetOrId: ruleset });
+      expect(rootNodes.length).to.eq(2);
+
+      const descriptor = await Presentation.presentation.getNodesDescriptor({ imodel: testIModel, rulesetOrId: ruleset, parentKey: rootNodes[0].key });
+      // ensure descriptor returns a correct ruleset
+      expect(descriptor!.hierarchyLevelRuleset!.rules).to.deep.equal([{
```

Comment:
you could avoid `descriptor!` by adding `assert(descriptor)` after retrieving it

Reply:
added